src/realm/node_header.hpp

```diff
@@ -229,16 +303,472 @@ class NodeHeader {
+    REALM_ASSERT_DEBUG(get_encoding(header) == Encoding::Flex);
+    REALM_ASSERT_DEBUG(bits_per_element <= 64);
+    REALM_ASSERT_DEBUG(bits_per_element > 0);
+    uint16_t word = (reinterpret_cast<uint16_t*>(header))[1];
+    word &= ~(0b111111 << 10);
+    //  we only have 6 bits, so store values in range 1-64 as 0-63
+    word |= (bits_per_element - 1) << 10;
+    (reinterpret_cast<uint16_t*>(header))[1] = word;
+}
+
+inline void NodeHeader::set_elementB_size(char* header, size_t bits_per_element)
+{
+    // we're a bit low on bits for the Flex encoding, so we need to squeeze stuff
+    REALM_ASSERT_DEBUG(get_encoding(header) == Encoding::Flex);
+    REALM_ASSERT_DEBUG(bits_per_element <= 64);
+    REALM_ASSERT_DEBUG(bits_per_element > 0);
+    uint16_t word = (reinterpret_cast<uint16_t*>(header))[3];
+    word &= ~(0b111111 << 10);
+    //  we only have 6 bits, so store values in range 1-64 as 0-63
+    word |= (bits_per_element - 1) << 10;
+    (reinterpret_cast<uint16_t*>(header))[3] = word;
+}
+
+inline size_t NodeHeader::get_elementA_size(const char* header)
+{
+    const auto encoding = get_encoding(header);
+    REALM_ASSERT_DEBUG(encoding == Encoding::Flex);
+    uint16_t word = (reinterpret_cast<const uint16_t*>(header))[1];
+    auto bits_per_element = (word >> 10) & 0b111111;
+    //  we only have 6 bits, so store values in range 1-64 as 0-63
+    // this means that Flex cannot support element sizes of 0
+    bits_per_element++;
+    REALM_ASSERT_DEBUG(bits_per_element <= 64);
+    REALM_ASSERT_DEBUG(bits_per_element > 0);
+    return bits_per_element;
+}
+
+inline size_t NodeHeader::get_elementB_size(const char* header)
+{
+    REALM_ASSERT_DEBUG(get_encoding(header) == Encoding::Flex);
+    uint16_t word = (reinterpret_cast<const uint16_t*>(header))[3];
+    auto bits_per_element = (word >> 10) & 0b111111;
+    // same as above
+    bits_per_element++;
+    REALM_ASSERT_DEBUG(bits_per_element <= 64);
+    REALM_ASSERT_DEBUG(bits_per_element > 0);
+    return bits_per_element;
+}
+
+inline size_t NodeHeader::get_num_elements(const char* header, Encoding encoding)
+{
+    switch (encoding) {
+        case NodeHeader::Encoding::Packed: {
+            REALM_ASSERT_DEBUG(get_encoding(header) == Encoding::Packed);
+            return (reinterpret_cast<const uint16_t*>(header))[3];
+        } break;
+        case NodeHeader::Encoding::WTypBits: {
+            REALM_ASSERT_DEBUG(get_wtype_from_header(header) == wtype_Bits);
+            return NodeHeader::get_size_from_header(header);
+        } break;
+        case NodeHeader::Encoding::WTypMult: {
+            REALM_ASSERT_DEBUG(get_wtype_from_header(header) == wtype_Multiply);
+            return NodeHeader::get_size_from_header(header);
+        } break;
+        case NodeHeader::Encoding::WTypIgn: {
+            REALM_ASSERT_DEBUG(get_wtype_from_header(header) == wtype_Ignore);
+            return NodeHeader::get_size_from_header(header);
+        } break;
+        default:
```

Comment:
How about Encoding::Flex? Confusing.

Reply:
The user of Encoding::Flex should not be calling this method, since it uses A and B for storing values and indices, calling `NodeHeader::get_num_elements()` is an error if the array is encoded in Flex IMO.